Add new action to run UI test by Appium

yonekawa opened the pull request:
Added new action to run UI test by [Appium](http://appium.io/) with RSpec.

``` rb
appium(
  app_path:  'appium/apps/TargetApp.app',
  spec_path: 'appium/spec',
  platform:  'iOS',
  caps: {
    versionNumber: '9.1',
    deviceName:    'iPhone 6'
  }
)
```

This is spec example.

``` rb
describe 'Examples' do
  it 'should display success' do
    find_element(:name, 'Show').click
    expect(find_element(:name, 'Success').displayed?).to be_truthy
  end
end
```

See [this document](http://appium.io/getting-started.html?lang=en) for more detail.
Please review it :smiley:

yonekawa committed:
Add new action to run UI test by Appium

yonekawa commented:
@xfreebird Oh, right. I will add an option for `npm install appium`.
I have a two plan to take into consideration it.

1. `appium_executable_path`: Developer can specify appium executable path.
2. `use_npm_bin`: if true, use `npm bin` to path of appium command.

Which do you like?

KrauseFx commented on docs/Actions.md:
```diff
@@ -546,6 +546,22 @@ ensure_no_debug_code(text: "NSLog",
 
+### [Appium](http://appium.io/)
+
+Run UI testing by Appium::Driver with RSpec.
```
Small thing, but you might want to add \` around Appium::Driver

KrauseFx commented on docs/Actions.md:
````diff
@@ -546,6 +546,22 @@ ensure_no_debug_code(text: "NSLog",
+
+```ruby
+appium(
+  app_path:  'appium/apps/TargetApp.app',
````
Can you change the quotes to `"` here in the `Actions.md`?

KrauseFx commented on fastlane.gemspec:
```diff
@@ -35,6 +35,9 @@ Gem::Specification.new do |spec|
   spec.add_dependency 'credentials_manager', '>= 0.11.0', '< 1.0.0' # Password Manager
   spec.add_dependency 'spaceship', '>= 0.14.2', '< 1.0.0' # communication layer with Apple's web services
 
+  spec.add_dependency 'appium_lib', '>= 4.1.0' # UI testing with Appium
```
We can't add new dependencies just for new actions. Instead take a look at the `cocoapods.rb` action: basically you run `verify_gem!("name")` to make sure a certain gem is installed. The user has to add the gem to their `Gemfile`

KrauseFx commented on fastlane.gemspec:
```diff
@@ -35,6 +35,9 @@ Gem::Specification.new do |spec|
   spec.add_dependency 'spaceship', '>= 0.14.2', '< 1.0.0' # communication layer with Apple's web services
 
+  spec.add_dependency 'appium_lib', '>= 4.1.0' # UI testing with Appium
+  spec.add_dependency 'rspec', '~> 3.1.0' # Development and Appium spec
```
Same as above, you can move this back to the `Development` section :+1:

KrauseFx commented on lib/fastlane/actions/appium.rb:
```diff
@@ -0,0 +1,145 @@
+        Helper.log.info "Parameter Invoke Appium Server: #{params[:invoke_appium_server]}"
+        Helper.log.info "Parameter Appium Server Host: #{params[:host]}"
+        Helper.log.info "Parameter Appium Server Port: #{params[:port]}"
+        Helper.log.info "Parameter Appium Capabilities: #{params[:caps]}"
```
I haven't tried it, but you could use

```ruby
FastlaneCore::PrintTable.print_values(config: params,
                                         hide_keys: [:whatever_here, :so_optional],
                                             title: "Summary for Appium Action")
```

KrauseFx commented on lib/fastlane/actions/appium.rb:
```diff
@@ -0,0 +1,145 @@
+        end
+
+        fork do
+          Process.exec("appium -a #{params[:host]} -p #{params[:port]}")
```
Is this necessary? Why didn't you use `sh` instead?

yonekawa replied:
`appium` command invokes HTTP server process.
Appium server is not killed at action finished when I use `sh`.
I would like to kill `appium` process at action finished for idempotence.

Do you have any ideas?

KrauseFx commented on lib/fastlane/actions/appium.rb:
```diff
@@ -0,0 +1,145 @@
+          FastlaneCore::ConfigItem.new(
+            key: :spec_path,
+            env_name: 'FL_APPIUM_SPEC_PATH',
+            description: 'Appium spec directory path',
```
I don't use Appium myself, but aren't there some default values? If for example there is always a `spec` folder, you can add

```
Dir["./spec"].last
```

which will either be nil (if not there) or a path (if available)

yonekawa replied:
Unfortunately, Appium does not have rule of directory structure such  as `spec/`.
I think that user user would like to specify directory flexibly.

KrauseFx commented on spec/actions_specs/appium_spec.rb:
```diff
@@ -0,0 +1,147 @@
+                spec_path: 'appium/spec'
+              })
+            end").runner.execute(:test)
+          end.to raise_error
```
Wouldn't it make sense to specify the exception message here to?

KrauseFx commented:
Hey @yonekawa, thanks for this great pull request. I added some comment before it's ready to merge, let me know once it's ready for review :+1:

xfreebird commented:
@yonekawa I would keep it simple and let the action detect the path.
I would use ```appium_path``` that by default would look first for the existence of ```/usr/local/bin/appium``` first and then for ```/Applications/Appium.app```.

If specified, it would check if the path ends in ```.app``` then it would use the bundled ```appium```, otherwise it would be the direct path to the ```appium``` (e.g. ```/usr/local/bin/appium```)

yonekawa commented:
@KrauseFx 
Fixed for your review. Could you review again?
And I to replied some comment. please confirm these.

@xfreebird 
I added `appium_path` option and a method to detecting executable.
Could you confirm it?

yonekawa commented:
@KrauseFx @xfreebird 
I'm looking forward to review :smiley:

KrauseFx commented:
Sorry for the late reply, reviewing your pull request right now :+1:

KrauseFx commented on fastlane.gemspec:
```diff
@@ -57,4 +57,5 @@ Gem::Specification.new do |spec|
   spec.add_development_dependency 'webmock', '~> 1.19.0'
   spec.add_development_dependency 'coveralls'
   spec.add_development_dependency 'rubocop', '~> 0.29'
+  spec.add_development_dependency 'appium_lib', '~> 4.1.0'
```
Please don't add this as a dependency, but add the following line to the top of your `self.run` method instead:

```ruby
Actions.verify_gem!('appium_lib')
```

This way the dependency doesn't get installed for every user, but only for people who actually want to use this action :+1:

KrauseFx replied:
Oh you already have that, so it's enough to just remove the dependency then.

yonekawa replied:
I tried to remove it from dependency. but spec failed because I wrote `require 'appium_lib'`
https://github.com/fastlane/fastlane/pull/935/files#diff-d3e4f0808df6e05aab7fce9a3a6cf589R14

Do you have any idea to fix this?

yonekawa commented:
@KrauseFx 
I replied to your two comments. please confirm it.

https://github.com/fastlane/fastlane/pull/935#discussion_r46376401
https://github.com/fastlane/fastlane/pull/935#discussion_r46376329

mfurtak commented:
@snatchev Do you have any thoughts on how best to resolve the `require` for the spec?

KrauseFx commented:
Hey everyone, just took a look, the dependency is fine, as it's only a development dependency. That means it's only installed on our local machines and for tests in general, but not for the end-user.

KrauseFx commented:
I haven't tried the action, but from the code it looks good to go for me. What do you think @mfurtak? :+1:

mfurtak commented:
I think the code looks great! Nice work @yonekawa :raised_hands: If you can squash it down to one commit, then I think we can ship it! 🎊

yonekawa commented:
@mfurtak Thank you! I squashed it down to one commit.

mfurtak commented:
@yonekawa こちらこそありがとうございます :rocket: :sparkles:

alexprice91 commented:
Any thoughts on how to get this to work with the `scan` action? Really do no want to use xcodebuild and then run this
